### PR TITLE
Split PictureLoader out of CardDatabase

### PR DESCRIPTION
**Pull request by ctrlaltca**
This PR moves all the card picture loading code out of carddatabase.cpp and into a new pictureloader.cpp file. This should make maintenance easier.
Also, Hq images handling has been removed; it was already disabled previously, and half hacked into being used for “custom” urls. Instead, give users the possibility to define both custom urls, primary (cards with muid) and fallback (cards without muid).

This is basically #1556, minus the bits causing problems.

### CODE REVIEW

File: cockatrice/src/pictureloader.cpp [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+#include <QNetworkReply>
+#include <QNetworkRequest>
+#include <QPainter>
+#include <QPixmapCache>
+#include <QSet>
+#include <QSvgRenderer>
+#include <QThread>
+#include <QUrl>
+
+class PictureToLoad::EnabledAndKeyCompareFunctor {
```
Comment (Daenyth) [inline cockatrice/src/pictureloader.cpp]: I know it's C++ convention to call this Functor, but can we please not? It's extremely confusing when you're used to using actual functors (from category theory).

`EnabledAndKeyCompareFunction` would be fine by me..

File: cockatrice/src/pictureloader.cpp [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+public:
+    inline bool operator()(CardSet *a, CardSet *b) const
```
Comment (Daenyth) [inline cockatrice/src/pictureloader.cpp]: A little function doc here describing the usage would help. Something like

> Returns true if `a` has higher download priority than b

File: cockatrice/src/pictureloader.cpp (hunk 1 above)
Comment (Daenyth) [inline cockatrice/src/pictureloader.cpp]: Actually why don't we call it `setDownloadPriorityComparator`

File: cockatrice/src/pictureloader.cpp [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+PictureToLoad::PictureToLoad(CardInfo *_card)
+    : card(_card), setIndex(0)
+{
+    if (card) {
+        sortedSets = card->getSets();
+        qSort(sortedSets.begin(), sortedSets.end(), EnabledAndKeyCompareFunctor());
+    }
+}
+
+bool PictureToLoad::nextSet()
```
Comment (Daenyth) [inline cockatrice/src/pictureloader.cpp]: Can we do `hasNextSet`?

File: cockatrice/src/pictureloader.cpp [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+                        << picsPath + "/downloadedPics/" + setName + "/" + correctedCardname;
+        }
+
+        QImage image;
+        QImageReader imgReader;
+        imgReader.setDecideFormatFromContent(true);
+        bool found = false;
+
+        //Iterates through the list of paths, searching for images with the desired name with any QImageReader-supported extension
+        for (int i = 0; i < picsPaths.length() && !found; i ++) {
```
Comment (Daenyth) [inline cockatrice/src/pictureloader.cpp]: I think it might make this more readable/maintainable to extract the logic out to a `bool cardImageExistsOnDisk` function (or maybe return `QString`/`nullptr` as result). Then in this function you can just `if cardImageExistsOnDisk(card, set)`

File: cockatrice/src/pictureloader.cpp [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+        QUrl redirectUrl = reply->attribute(QNetworkRequest::RedirectionTargetAttribute).toUrl();
+        QNetworkRequest req(redirectUrl);
+        qDebug() << "following redirect:" << cardBeingDownloaded.getCard()->getName() << "Url:" << req.url();
+        networkManager->get(req);
+        return;
+    }
+
+    const QByteArray &picData = reply->peek(reply->size()); //peek is used to keep the data in the buffer for use by QImageReader
+
+    // check if the image is blacklisted
```
Comment (Daenyth) [inline cockatrice/src/pictureloader.cpp]: Rather than the comment I'd extract out to a method and `if imageIsBlackListed(picData)`

File: cockatrice/src/pictureloader.cpp [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+    } else {
+        QPixmapCache::insert(card->getPixmapCacheKey(), QPixmap::fromImage(image));
+    }
+
+    card->emitPixmapUpdated();
+}
+
+void PictureLoader::clearPixmapCache(CardInfo *card)
+{
+    //qDebug() << "Deleting pixmap for" << name;
```
Comment (Daenyth) [inline cockatrice/src/pictureloader.cpp]: I'd either uncomment this or delete the line.

File: cockatrice/src/pictureloader.cpp [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+
+void PictureLoader::clearPixmapCache()
+{
+    QPixmapCache::clear();
+}
+
+void PictureLoader::cacheCardPixmaps(QList<CardInfo *> cards)
+{
+    QPixmap tmp;
+    // never cache more than 300 cards at once for a single deck
```
Comment (Daenyth) [inline cockatrice/src/pictureloader.cpp]: We should pull out a constant here.

File: doc/cards.xsd [hunk 8]
```diff
@@ -29,12 +29,12 @@
                           <xs:extension base="xs:string">
                             <xs:attribute type="xs:int" name="muId" use="optional"/>
                             <xs:attribute type="xs:anyURI" name="picUrl" use="optional"/>
-                            <xs:attribute type="xs:anyURI" name="picUrlHq" use="optional"/>
```
Comment (Daenyth) [inline doc/cards.xsd]: Why remove this? It's still valid, no? Just because we don't have a source right now doesn't mean someone might not want to supply one manually.

File: oracle/src/oracleimporter.h [hunk 9]
```diff
@@ -2,6 +2,7 @@
 #define ORACLEIMPORTER_H
 
 #include <QMap>
+#include <QVariant>
```
Comment (Daenyth) [inline oracle/src/oracleimporter.h]: Why is this needed now if no other code changed in the file?

**Comment by Daenyth** [comment]
This would be a great PR to start using googletest with, though I'd be ok with leaving it out.

We should _heavily_ test this before merge as it's pretty core behavior.

File: cockatrice/src/pictureloader.cpp (hunk 3 above)
Comment (ctrlaltca) [inline cockatrice/src/pictureloader.cpp]: It's not exactly a ```bool hasNextSet()```, because it also increments the internal index to point to the next set (i'd expect a hasX() method to be object const).

File: doc/cards.xsd (hunk 8 above)
Comment (ctrlaltca) [inline doc/cards.xsd]: Changes included in this PR removes che choice between SQ and HQ images; you can get them form the "default source" (currently, the gatherer) or a custom source.

File: oracle/src/oracleimporter.h (hunk 9 above)
Comment (ctrlaltca) [inline oracle/src/oracleimporter.h]: Before, ``` #include <carddatabase.h>``` would get in some includes for Qt objects that included <QVariant> themselves. With this PR a lot of includes have been removed from carddatabase.h, and i had to add this to fix compilation.

**Comment by ctrlaltca** [comment]
Anyone caring to test?

**Comment by ZeldaZach** [comment]
This seems to work fine to me.

I tested it through altering URLs, Making tokens (of real cards, tokens, and fake cards that don't exist), and trying to modify the game play. No crashes of any kind detected at this point.

When altering the URLs, if the first fails it goes to the secondary URL (as it should). When both are invalid we really should default to SOMEWHERE (probably gatherer).

I like it, but I'd like to see another test to confirm. :shipit:

**Comment by Daenyth** [comment]
I'll try to look at this and signoff soon
